test/test.js

```diff
@@ -151,6 +151,35 @@ test('create: applies data to templated element nodes', () => {
   is(root.innerHTML, `<div>Hello world</div>`)
 })
 
+test('create: applies data to templated DocumentFragment nodes', () => {
+  // https://github.com/github/template-parts/pull/73
```

Comment:
I read the [section in the project's README](https://github.com/dy/template-parts?tab=readme-ov-file#alternatives) about alternatives, and was surprised to see these PRs mentioned explicitly. As the author of 65, 72, and 73, I'm curious about this project's goals.

Are you hoping to achieve parity with the GitHub implementation? What are the barriers you are facing to upstreaming the parts of this project that have outpaced the GitHub version?

Are you open to porting the `InnerTemplatePart` changes to the GitHub project? Are you hoping to remain separate?

Reply:
@seanpdoyle first of all thank you for contributing to this proposal, I see your PRs to the main repo.
I wasn't sure what would be the proper behaviour for nested templates. Is that in standard? Maybe you have practical scenario when you need internal template rendered right away? I guess that would be simpler which is good and more intuitive, but what if we need to have internal templates intact, is that possible scenario? How would we approach it?
`<template>` is sometimes used in alternative sense, like `<template shadowrootmode="open"></template>` or `<template x-for="i in 10"><li x-text="i"></li></template>` in alpine/sprae/alike.
As for the repository - I intended it for upstream first, but then figured out I din't like the pipeline/build scripts, also the amount of changes was a bit overwhelming (almost everything has changed!), so opted for a fork.
It would definitely be good to be on par with github, as I consider it main implementation. I think I am collecting changes and one day will patch them here.

Reply:
I'm hoping to make something like what's outlined in the [`seanpdoyle/turbo_stream_button` README](https://github.com/seanpdoyle/turbo_stream_button?tab=readme-ov-file#advanced-append-form-controls) more straightforward to implement:

```html
<div id="to-be-updated">To be updated</div>

<form>
  <template>
    <turbo-stream action="update" target="to-be-updated">
      <template>Updated with: {{value}}</template>
    </turbo-stream>
  </template>

  <label for="value">Value</label>
  <input id="value" name="value">

  <button>Update</button>
</form>

<script>
  addEventListener("submit", (event) => {
    event.preventDefault()
  
    const form = event.target
    const formData = new FormData(form)
    const properties = Object.fromEntries(formData.entries())
    const templates = form.querySelectorAll("template")
    
    const templateInstances = Array.from(templates, template => new TemplateInstance(template, properties))

    templateInstances.forEach(instance => form.append(instance))
  })
</script>
```

Without this change, the `{ value }` derived from the `FormData` instance isn't available to the `turbo-stream template` element.

> I wasn't sure what would be the proper behaviour for nested templates. Is that in standard?

While I understand that inner templates can have attributes that warrant special treatment (like `[directive]` and `[expression]`), I guarded this PR's changes with a check for those attributes. When absent, treat the inner `<template>` content as an extension of the outer `<template>` content.

I don't see a comparable scenario in the specification, so I'm not sure how to interpret that.

Reply:
The specification outlines [a hypothetical `foreach` directive](https://github.com/WICG/webcomponents/blob/159b1600bab02fe9cd794825440a98537d53b389/proposals/Template-Instantiation.md#33-conditionals-and-loops-using-nested-templates) which is executed with its own scope of properties. It also mentions an `if` expression, but doesn't share any (pseudo-)code.

The lack of clarity makes it unclear whether or not outer context variables should be accessible from inner contexts.

As a consumer of the interface, I would expect both `value` and `showValue` to be accessible in the inner `<template>` element. For example, consider something like:

```html
<template id="outer">
  <template directive="if" expression="showValue">{{value}}</template>
</template>

<script>
  const template = document.getElementById("outer")
  const instance = new TemplateInstance(template, {
    value: "visible",
    showValue: true
  })

  document.body.append(instance) // should have text "visible"

  instance.update({ showValue: false }) // should no longer have text "visible"
</script>
```